#' Return a genotype matrix based on risk variants from a population
#' @param popfilename The (binary, gzipped) file containing a population simulated by TFL2013_ind
#' @param popfileOffset The offset (in bytes) of the relevant record in popfilename
#' @param phenofilename The (binary, gzipped) file containing phenotypes generated by TFL2013_ind (optional)
#' @param phenofileOffset The offset (in bytes) of the relevant record in phenofilename (optional)
#' @param modelName The model of gene action.  Must be one of: recessive, additive, multiplicative, or popgen
#' @param dominance The dominance of a risk mutation.  Only applies to the "popgen" model of gene action
#' @return traits, which is a list of trait (or genetic) values for each individual
#' @return genos, which is an data frame of genotypes. Rows = individuals. Columns = 0,1,2 copies of risk mutation
#' @return nremoved, which is the number of markers removed because they resulted in duplicate columns in genos
#' @details
#' The order of the columns in "genos" is in descending order of both frequency and absolute value of effect size
#' If phenofilename and phenofileOffset are both provided, then "traits" corresponds to individual phenotypes.
#' If these two arguments are NOT provided, then "traits" corresponds to the genetic component of trait value,
#' which is calculated by the parameter bassed to modelName.
#' Risk variant frequencies are calculated by colSums(genos)/(2*nrow(genos)), in case you want to filter
getRiskVariantMatrix = function(popfilename,
    popfileOffset,
    phenofilename,
    phenofileOffset,
    modelName = "recessive",
    dominance = 0)
    {
        if( missing(popfilename) ) {
            stop("Error: popfilename required")
        }
        if( missing(popfileOffset) ) {
            stop("Error: popfileOffset required")
        }
        if( popfileOffset < 0 ) {
            stop("Error: popfileOffset must be >= 0")
        }
        if(modelName != "recessive" &
           modelName != "additive" &
           modelName != "multiplicative" &
           modelName != "popgen")
            {
                stop("Error: invalid model name.  Must be one of: recessive, additive, multiplicative, or popgen")
            }
        if( missing(phenofilename) )
            {
                XX = .getRiskVariantMatrixDetails(modelName,popfilename,popfileOffset,dominance)
                return (XX);
            }
            
        if( missing(phenofileOffset) )
            {
                stop("Error: offest missing for phenotype file");
            }
        XX=.getRiskVariantMatrixDetailsPheno(modelName,popfilename,popfileOffset,phenofilename,phenofileOffset,dominance)
        return (XX);
    }
